cupyx/scipy/linalg/special_matrices.py

```diff
@@ -0,0 +1,552 @@
+        arrs = cupy.atleast_2d(*arrs)
+    if any(a.ndim != 2 for a in arrs):
+        bad = [k for k in range(len(arrs)) if arrs[k].ndim != 2]
+        raise ValueError("arguments in the following positions have dimension "
+                         "greater than 2: {}".format(bad))
+
+    shapes = tuple(a.shape for a in arrs)
+    shape = tuple(sum(x) for x in zip(*shapes))
+    dtype = cupy.find_common_type([a.dtype for a in arrs], [])
+    out = cupy.zeros(shape, dtype=dtype)
+    r, c = 0, 0
+    for arr in arrs:
+        rr, cc = arr.shape
+        out[r:r + rr, c:c + cc] = arr
+        r += rr
+        c += cc
+    return out
+
+
+def companion(a):
+    """Create a companion matrix.
+
+    Create the companion matrix associated with the polynomial whose
+    coefficients are given in ``a``.
+
+    Args:
+        a (cupy.ndarray): 1-D array of polynomial coefficients. The length of
+            ``a`` must be at least two, and ``a[0]`` must not be zero.
+
+    Returns:
+        (cupy.ndarray): The first row of the output is ``-a[1:]/a[0]``, and the
+            first sub-diagonal is all ones. The data-type of the array is the
+            same as the data-type of ``-a[1:]/a[0]``.
+
+    .. seealso:: :func:`cupyx.scipy.linalg.fielder_companion`
+    .. seealso:: :func:`scipy.linalg.companion`
+    """
+    n = a.size
+    if a.ndim != 1:
+        raise ValueError("`a` must be one-dimensional.")
+    if n < 2:
+        raise ValueError("The length of `a` must be at least 2.")
+    if a[0] == 0:
+        raise ValueError("The first coefficient in `a` must not be zero.")
+    first_row = -a[1:] / a[0]
+    c = cupy.zeros((n - 1, n - 1), dtype=first_row.dtype)
+    c[0] = first_row
+    cupy.fill_diagonal(c[1:], 1)
+    return c
+
+
+def helmert(n, full=False):
+    """Create an Helmert matrix of order ``n``.
+
+    This has applications in statistics, compositional or simplicial analysis,
+    and in Aitchison geometry.
+
+    Args:
+        n (int): The size of the array to create.
+        full (bool, optional): If True the (n, n) ndarray will be returned.
+            Otherwise, the default, the submatrix that does not include the
+            first row will be returned.
+
+    Returns:
+        cupy.ndarray: The Helmert matrix. The shape is (n, n) or (n-1, n)
+            depending on the ``full`` argument.
+
+    .. seealso:: :func:`scipy.linalg.helmert`
+    """
+    d = cupy.arange(n)
+    H = cupy.tri(n, n, -1)
+    H.diagonal()[:] -= d
+    d *= cupy.arange(1, n+1)
+    H[0] = 1
+    d[0] = n
+    H /= cupy.sqrt(d)[:, None]
+    return H if full else H[1:]
+
+
+def hilbert(n):
+    """Create a Hilbert matrix of order ``n``.
+
+    Returns the ``n`` by ``n`` array with entries ``h[i,j] = 1 / (i + j + 1)``.
+
+    Args:
+        n (int): The size of the array to create.
+
+    Returns:
+        cupy.ndarray: The Hilbert matrix.
+
+    .. seealso:: :func:`scipy.linalg.hilbert`
+    """
+    values = cupy.arange(1, 2*n, dtype=float)
```

Comment:
Please use `cupy.float64` for the readability.
```suggestion
    values = cupy.arange(1, 2*n, dtype=numpy.float64)
```